File: stripe/src/main/java/com/stripe/android/model/AccountParams.java
package com.stripe.android.model;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

/**
 * Represents a grouping of parameters needed to create a Token for a Connect account on the server.
 */
public final class AccountParams implements StripeParamsModel {

    static final String API_BUSINESS_TYPE = "business_type";
    static final String API_TOS_SHOWN_AND_ACCEPTED = "tos_shown_and_accepted";

    private final boolean mTosShownAndAccepted;
    @Nullable private final BusinessType mBusinessType;
    @Nullable private final Map<String, Object> mBusinessData;

    /**
     * Create an {@link AccountParams} instance for a {@link BusinessType#Individual} or
     * {@link BusinessType#Company}.
     *
     * @param tosShownAndAccepted Whether the user described by the data in the token has been shown
     *                            the <a href="https://stripe.com/docs/api/tokens/create_account#create_account_token-account-tos_shown_and_accepted">Stripe Connected Account Agreement</a>.
     *                            When creating an account token to create a new Connect account,
     *                            this value must be <code>true</code>.
     * @param businessType        See {@link BusinessType}
     * @param businessData        A map of <a href="https://stripe.com/docs/api/accounts/create#create_account-company">company</a>
     *                            or <a href="https://stripe.com/docs/api/accounts/create#create_account-individual">individual</a> params.
     *
     *
     * @return {@link AccountParams}
     */
    @NonNull
    public static AccountParams createAccountParams(
            boolean tosShownAndAccepted,
            @Nullable BusinessType businessType,
            @Nullable Map<String, Object> businessData) {
        return new AccountParams(businessType, businessData, tosShownAndAccepted);
    }

    private AccountParams(@Nullable BusinessType businessType,
                          @Nullable Map<String, Object> businessData,
                          boolean tosShownAndAccepted) {
        mBusinessType = businessType;
        mBusinessData = businessData;
        mTosShownAndAccepted = tosShownAndAccepted;
    }

    /**
     * Create a string-keyed map representing this object that is ready to be sent over the network.
     *
     * @return a String-keyed map
     */
    @NonNull
    @Override
    public Map<String, Object> toParamMap() {
        final Map<String, Object> accountData = new HashMap<>();
        if (mBusinessType != null) {
            accountData.put(API_BUSINESS_TYPE, mBusinessType.code);

            if (mBusinessData != null) {
                accountData.put(mBusinessType.code, mBusinessData);
            }
        }
        accountData.put(API_TOS_SHOWN_AND_ACCEPTED, mTosShownAndAccepted);

        final Map<String, Object> params = new HashMap<>();
        params.put("account", accountData);
        return params;
    }

    @Override
    public int hashCode() {
        return Objects.hash(mTosShownAndAccepted, mBusinessType, mBusinessData);
    }

    @Override
    public boolean equals(@Nullable Object obj) {
        return this == obj || (obj instanceof AccountParams && typedEquals((AccountParams) obj));
    }

    private boolean typedEquals(@NonNull AccountParams accountParams) {
        return Objects.equals(mTosShownAndAccepted, accountParams.mTosShownAndAccepted)
                && Objects.equals(mBusinessType, accountParams.mBusinessType)
                && Objects.equals(mBusinessData, accountParams.mBusinessData);
    }

    /**
     * See <a href="https://stripe.com/docs/api/accounts/create#create_account-business_type">
     *     Account creation API docs</a>
     */
    public enum BusinessType {
        Individual("individual"),
        Company("company");

        @NonNull public final String code;

        BusinessType(@NonNull String code) {
            this.code = code;
        }
    }
}

File: stripe/src/main/java/com/stripe/android/model/AccountParams.kt
package com.stripe.android.model

import java.util.Objects

/**
 * Represents a grouping of parameters needed to create a Token for a Connect account on the server.
 */
class AccountParams private constructor(
    private val businessType: BusinessType?,
    private val businessData: Map<String, Any>?,
    private val tosShownAndAccepted: Boolean
) : StripeParamsModel {

    /**
     * Create a string-keyed map representing this object that is ready to be sent over the network.
     *
     * @return a String-keyed map
     */
    override fun toParamMap(): Map<String, Any> {
        val accountData =
            mapOf(API_TOS_SHOWN_AND_ACCEPTED to tosShownAndAccepted)

        val businessData =
            if (businessType != null) {
                mapOf(API_BUSINESS_TYPE to businessType.code)
                    .plus(
                        businessData?.let {
                            mapOf(businessType.code to businessData)
                        } ?: emptyMap()
                    )
            } else {
                emptyMap()
            }

        return mapOf("account" to accountData.plus(businessData))
    }

    override fun hashCode(): Int {
        return Objects.hash(tosShownAndAccepted, businessType, businessData)
    }

    override fun equals(other: Any?): Boolean {
        return when {
            this === other -> true
            other is AccountParams -> typedEquals(other)
            else -> false
        }
    }

    private fun typedEquals(accountParams: AccountParams): Boolean {
        return tosShownAndAccepted == accountParams.tosShownAndAccepted &&
            businessType == accountParams.businessType &&
            businessData == accountParams.businessData
    }

    /**
     * See [Account creation API docs](https://stripe.com/docs/api/accounts/create#create_account-business_type)
     */
    enum class BusinessType constructor(val code: String) {
        Individual("individual"),
        Company("company")
    }

    companion object {
        internal const val API_BUSINESS_TYPE = "business_type"
        internal const val API_TOS_SHOWN_AND_ACCEPTED = "tos_shown_and_accepted"

        /**
         * Create an [AccountParams] instance for a [BusinessType.Individual] or
         * [BusinessType.Company].
         *
         * @param tosShownAndAccepted Whether the user described by the data in the token has been shown
         * the [Stripe Connected Account Agreement](https://stripe.com/docs/api/tokens/create_account#create_account_token-account-tos_shown_and_accepted).
         * When creating an account token to create a new Connect account,
         * this value must be `true`.
         * @param businessType See [BusinessType]
         * @param businessData A map of [company](https://stripe.com/docs/api/accounts/create#create_account-company)
         * or [individual](https://stripe.com/docs/api/accounts/create#create_account-individual) params.
         *
         *
         * @return [AccountParams]
         */
        @JvmStatic
        fun createAccountParams(
            tosShownAndAccepted: Boolean,
            businessType: BusinessType?,
            businessData: Map<String, Any>?
        ): AccountParams {
            return AccountParams(businessType, businessData, tosShownAndAccepted)
        }
    }
}
